### ISSUE TITLE
Question about Table3

### ISSUE THREAD

**Issue by j84637587** [firework8/ProtoGCN#3]
In Table 3, you present an ablation study analyzing the contribution of each component in ProtoGCN under the NTU-120 X-Sub setting.

The question arises: without CSCL, does MTE and PRN have no role in improving the performance(Only Cross Entropy)? In other words, do MTE and PRN only contribute when CSCL is included?

If this is the case, in rows 1, 2, and 4 of the table, the accuracy values would likely remain the same or show minimal variation?

![Image](https://github.com/user-attachments/assets/ba6270cf-c161-4cd4-9b18-421a851cfc14)

**Comment by firework8** [comment]
Thanks for your attention. The presentation of Table 3 here actually misses a note.
In practice, the configuration of rows 1, 2, and 4 in Table 3 is supervised by using the Cross Entropy loss, rather than the contrastive objective of CSCL.

There is an interesting observation: MTE and PRN are valuable, however, the corresponding boosts are more significant under the CSCL constraints. 

Specifically, under normal CE loss, the changes in rows 1, 2, and 4 reflect that the results are not simply performance fluctuations. When CSCL is not included, MTE and PRN also contribute to performance. 

Furthermore, as can be seen from the result in row 3, the boost from CSCL itself is not significant. However, the combination of CSCL and PRN provides better results. This suggests that compared to CE loss, the contrastive constraints of CSCL can better help PRN to fulfill its capability. In other words, CSCL could result in more discriminative features after prototype reconstruction, but in which MTE and PRN also have an important role.

Hope this could help you.

**Comment by j84637587** [comment]
This is truly fascinating. 
It makes me realize that even if the loss computation path does not include MTE and PRN, MTE and PRN still influences the gradient computation, thereby improving the model.

But this also makes me curious about how their gradients are computed.  

Thank you for your reply.

**Comment by firework8** [comment]
Thanks for your approval.

In the standard case, MTE and PRN are updated by back propagation of CSCL to improve the GCN model.
Instead, when CSCL is not included, we replace L_CSC with L_CE for the computational path in the bottom part of Figure 2, while the computational path in the top part remains unchanged. In this case, the parameters of MTE and PRN are updated by back propagation of the newly added L_CE.

The gradient calculation is similar in both cases, except that it is switched from contrastive loss to cross-entropy loss.
You could also observe the change of gradient calculation by printing out the corresponding values in [base.py#L72](https://github.com/firework8/ProtoGCN/blob/main/protogcn/models/heads/base.py#L72).

Hope this could help you understand.

**Comment by firework8** [comment]
Specifically, the processing is as follows:

First, the output shape of PRN is V^2, C.
Then, in [protogcn.py#L187](https://github.com/firework8/ProtoGCN/blob/main/protogcn/models/gcns/protogcn.py#L187), it changes to V^2 through `mean()`.
Finally, in [Class_Specific_Contrastive_Loss.py#L89](https://github.com/firework8/ProtoGCN/blob/main/protogcn/models/losses/Class_Specific_Contrastive_Loss.py#L89), for CSCL, it becomes f_c.
For the replaced L_CE, the processing is the same, except that the function is switched from contrastive loss to cross-entropy loss.

Hope this could help you.

**Comment by j84637587** [comment]
Got it!
Thank you for your explanation!
